src/mi/Gaddress_validator.c

```diff
@@ -0,0 +1,349 @@
+/*
+ * Contributed by Stephen M. Webb <stephen.webb@bregmasoft.ca>
+ *
+ * This file is part of libunwind, a platform-independent unwind library.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "libunwind_i.h"
+
+#ifdef UNW_REMOTE_ONLY
+bool
+unw_address_is_valid(UNUSED unw_word_t addr, UNUSED size_t len)
+{
+  Debug(1, "remote-only invoked\n");
+  return false;
+}
+
+#else /* !UNW_REMOTE_ONLY */
+
+static pthread_once_t _unw_address_validator_init_once = PTHREAD_ONCE_INIT;
```

Comment:
how AS-safe is this, or how can we ensure, that this is called before being in a signal context?

Reply:
Do you mean the initialization of the variable? It's done by the dynamic loader when the RW LOAD segment is loaded. I'm going to assume the dynamic loader implements dlopen() (or whatever internally implements loading an ELF file on your favourite OS) in a threadsafe manner that prevents any code from being executed before it's loaded.

If you're asking about `pthread_run_once()`, well, it's threadsafe. It doesn't matter if the thread is executing in a signal context or not.

Reply:
The only time there would be a problem is if the first thread to run the `pthread_once()` function receives a signal while *in* that function and calls the `pthread_once()` function a second time from within the signal handler. That would potentially result in a deadlock -- in the case of GNU/Linux, it would spin forever on a futex.

Of course, the same problem already applies to `unw_init_local()` except it's far more likely to occur. My first idea was to add the initialization to `mi_init()` since that's protected by the mutex in `tdep_init()` but that quickly became a more invasive change for little benefit.

Other than just documenting the potential deadlock, can you suggest a better approach to initialization?

Reply:
> The only time there would be a problem is if the first thread to run the pthread_once() function receives a signal while in that function and calls the pthread_once() function

No, thats not what I have observed, when it comes to AS-unsafe functions. Most of the time, the thread does some libc stuff which in the end calls into the malloc part, and this takes a lock. If a signal is happening while holding the lock and the handler is calling into libunwind, it needs to avoid that it also calls something AS-unsafe which might also call into malloc which tries to take the same lock. Deadlock.

I don't know, if `pthread_once()` can call into `malloc()`, it does not look so but I cannot find something about AS-safe/unsafe for `pthread_once()` either.

The only thing I can comeup is, a documented way so, that all these things can be triggered voluntarily outside a signal handler. If this is just `unw_init_local()` its fine.

Reply:
Latest change replaces the pthread_once() with a simple atomic flag check that should be both thread safe and signal safe and definitely does not make a call to malloc().

Reply:
fine with me. thanks